django/contrib/postgres/forms/array.py

```diff
@@ -216,3 +216,23 @@ def clean(self, value):
         if errors:
             raise ValidationError(list(chain.from_iterable(errors)))
         return cleaned_data
+
+    def has_changed(self, initial, data):
+        try:
+            python_data = self.to_python(data)
+        except ValidationError:
+            pass
+        else:
+            if self.remove_trailing_nulls:
+                null_index = None
+                for i, value in reversed(list(enumerate(python_data))):
+                    if value in self.base_field.empty_values:
+                        null_index = i
+                    else:
+                        break
+                if null_index is not None:
+                    data = python_data[:null_index]
+
+        if initial in self.empty_values and data in self.empty_values:
+            return False
+        return super().has_changed(initial, data)
```

Comment:
I think we can add an internal hook `_remove_trailing_nulls()` to the `SplitArrayField` that can be used in `clean()` and `has_changed()`:
```python
    def _remove_trailing_nulls(self, values):
        errors = []
        null_index = None
        if self.remove_trailing_nulls:
            for i, value in reversed(list(enumerate(values))):
                if value in self.base_field.empty_values:
                    null_index = i
                else:
                    break
            if null_index is not None:
                values = values[:null_index]
        return values, null_index

    def has_changed(self, initial, data):
        data, _ = self._remove_trailing_nulls(self.to_python(data))
        return super().has_changed(initial, data)
```
Please add it in the 1st commit, the 2nd commit will contain a bugfix.

Reply:
This on its own will not fix the issue if `initial` is `None`. We also need to add:

```python
if initial is in self.empty_values and data is in self.empty_values:
   return False
```

Or something similar. At least I think, the assertion: `self.assertIs(field.has_changed(None, ['', '']), False)` is covering this case.

Reply:
Yes, we should use the same approach as in `SimpleArrayField`, e.g.
```python
    def has_changed(self, initial, data):
        try:
            data, _ = self._remove_trailing_nulls(self.to_python(data))
        except ValidationError:
            pass
        if initial in self.empty_values and data in self.empty_values:
            return False            
        return super().has_changed(initial, data)
```